internal/provisioner/plain/controllers/bundleinstance_controller.go

```diff
@@ -96,36 +103,50 @@ func (r *BundleInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		}
 	}()
 
-	b := &rukpakv1alpha1.Bundle{}
-	if err := r.Get(ctx, types.NamespacedName{Name: bi.Spec.BundleName}, b); err != nil {
-		bundleStatus := metav1.ConditionUnknown
-		if apierrors.IsNotFound(err) {
-			bundleStatus = metav1.ConditionFalse
+	bundle, oldBundles, err := r.reconcileDesiredBundle(ctx, bi)
+	if err != nil {
+		if errors.Is(err, ErrMaxGeneratedLimit) {
+			meta.SetStatusCondition(&bi.Status.Conditions, metav1.Condition{
+				Type:    rukpakv1alpha1.TypeHasValidBundle,
+				Status:  metav1.ConditionFalse,
+				Reason:  rukpakv1alpha1.ReasonMaxGeneratedBundlesReached,
+				Message: err.Error(),
+			})
+			return ctrl.Result{}, nil
 		}
 		meta.SetStatusCondition(&bi.Status.Conditions, metav1.Condition{
 			Type:    rukpakv1alpha1.TypeHasValidBundle,
-			Status:  bundleStatus,
-			Reason:  rukpakv1alpha1.ReasonBundleLookupFailed,
+			Status:  metav1.ConditionUnknown,
+			Reason:  rukpakv1alpha1.ReasonReconcileFailed,
 			Message: err.Error(),
 		})
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		return ctrl.Result{}, err
 	}
-
-	desiredObjects, err := r.loadBundle(ctx, bi)
-	if err != nil {
-		var bnuErr *errBundleNotUnpacked
-		if errors.As(err, &bnuErr) {
-			reason := fmt.Sprintf("BundleUnpack%s", b.Status.Phase)
-			if b.Status.Phase == rukpakv1alpha1.PhaseUnpacking {
-				reason = "BundleUnpackRunning"
-			}
-			meta.SetStatusCondition(&bi.Status.Conditions, metav1.Condition{
-				Type:   rukpakv1alpha1.TypeInstalled,
-				Status: metav1.ConditionFalse,
-				Reason: reason,
-			})
-			return ctrl.Result{}, nil
+	if bundle.Status.Phase != rukpakv1alpha1.PhaseUnpacked {
+		reason := rukpakv1alpha1.ReasonUnpackPending
+		status := metav1.ConditionTrue
+		message := fmt.Sprintf("Waiting for the %s Bundle to be unpacked", bundle.GetName())
```

Comment:
For this section where the bundle is not unpacked, would it make sense to grab most of this directly from the Bundle `Unpacked` condition and just bundle it up?

Would it make sense to get rid of the BI `HasValidBundle` condition and just take the latest bundle's `Unpacked` condition directly?

Reply:
> Would it make sense to get rid of the BI HasValidBundle condition and just take the latest bundle's Unpacked condition directly?

I'm not opposed to this suggestion but it's also a problem with the existing implementation so I'd like to hold off on changing this here.

Reply:
Created #333